src/msg/msg_types.h

```diff
@@ -541,12 +541,12 @@ WRITE_CLASS_ENCODER_FEATURES(entity_addr_t)
 
 std::ostream& operator<<(std::ostream& out, const entity_addr_t &addr);
 
-inline bool operator==(const entity_addr_t& a, const entity_addr_t& b) { return memcmp(&a, &b, sizeof(a)) == 0; }
-inline bool operator!=(const entity_addr_t& a, const entity_addr_t& b) { return memcmp(&a, &b, sizeof(a)) != 0; }
-inline bool operator<(const entity_addr_t& a, const entity_addr_t& b) { return memcmp(&a, &b, sizeof(a)) < 0; }
-inline bool operator<=(const entity_addr_t& a, const entity_addr_t& b) { return memcmp(&a, &b, sizeof(a)) <= 0; }
-inline bool operator>(const entity_addr_t& a, const entity_addr_t& b) { return memcmp(&a, &b, sizeof(a)) > 0; }
-inline bool operator>=(const entity_addr_t& a, const entity_addr_t& b) { return memcmp(&a, &b, sizeof(a)) >= 0; }
+inline bool operator==(const entity_addr_t& a, const entity_addr_t& b) { return a.nonce == b.nonce; }
```

Comment:
I don't think that comparing just `nonce` (while ignoring `type` and `u`) is correct. `nonce` should be randomized but, AFAIK, there is no uniqueness guarantee, especially among hosts. I recall a subtle bug in `crimson-osd` where all instances were assigned with the same, small integer (maybe `1` or `2`) but, generally, the comparator was still able to work correctly.

Reply:
Yeah, that was a rook deployment. the nonce value of different crimson instances are always identical despite that they are indeed *different*.

Reply:
Thank you Radoslaw and Kefu!
If I understand correctly, an entity `entity_addr_t` is identified by the following fields together: `nonce`, `type`, and `u`.
Direct use of the `memcmp` comparison on the `entity_addr_t` in the LE & BE machines leads to errors since of bytes sequence.

Can the `entity_addr_t` be compared like:
```
a.nonce == b.nonce &&
a.type == b.type &&
memcmp(&a.u, &b.u, sizeof(a.u)) == 0
```
where the `u` should be in the network order in both LE and BE?

Can only the `u` be enough to identify the instance?
Thank you!

Reply:
Dear Kefu,

Acomparison using memcmp should not work since of the LE/BE archs, due to the endianness and padding in the structs. I manage to try a direct comparison of the need fields (type, nonce, addr, port, + other info) like:

src/include/cmp.h

inline bool operator>=(const entity_addr_t& a, const entity_addr_t& b) {
  struct sockaddr_storage ssa = a.get_sockaddr_storage();
  struct sockaddr_storage ssb = b.get_sockaddr_storage();

  ssa.ss_family = htons(ssa.ss_family);
  ssb.ss_family = htons(ssb.ss_family);

  return memcmp(&ssa, &ssb, sizeof(ssa)) > 0 ||
         (memcmp(&ssa, &ssb, sizeof(ssa)) == 0 && (a.type > b.type ||
         (a.type == b.type && (a.nonce >= b.nonce))));
}

this should use the BE ints on both BE and LE nodes.

Could you please tell me which fields in the sockaddr may be needed, especially for the sockaddr_in6? I assume address and port as minimum.


Best regards,
Alex

Reply:
@aliakseimakarau hi Alex, i don't have a quick answer off top of my head. i'd suggest just the encoding scheme of `entity_addr_t` should be endianness-aware. let's take a closer look at muti-byte member variables of `entity_addr_t`, (assuming we only support post

- `type`: encoded as `le32`: safe
- `nonce`: encoded as `le32`: safe
- `elen`: encoded as `le32`: safe
- `u.sa.sa_family`: encoded as `le16`: safe
- `u.sa.sa_data[elen]`: encoded as opaque blob: not safe

so an IPv6 `sockaddr` looks like:
```c
sa_family_t family;
uint16_t port
uint32_t flowinfo;
uint32_t addr32[4];
uint32_t scope_id;
```

to the best of my knowledge, under most circumstances, we can assume an always-zero flowinfo and an always-zero scope_id in `sockaddr`. as the former is used for IPv6 flow labelling, and the latter is for identifying the scope of a link-local address or some special multicast address. neither of these use case of these scoped addressed should be used in a sane Ceph in production. but since the existing Ceph on-wire encoding schema always encodes them by stuffing the bufferlist with `u.sa.sa_data[elen]`. probably you could attack this problem in two steps:

1.  to ensure that the on-wire format of `entity_addr_t` is little-endian. and decode/encode this way on big-endian arch. 
2. if you want to avoid encoding `flowinfo` and `scope_id`, probably you need to add some safe guard at the places where we retrieve the socket address using `getsockname()` or `getifaddrs()` calls, and print a visible error when IPv6 is used *and* the `flowinfo` or `scope_id` is set for the address. 

i guess the 2nd step might be controversial, as some of us might argue that we should not encode this assumption in  our implementation, as it might be orthogonal to Ceph deployment. so we could have a different PR or commit for implementing this.

Reply:
Hi @tchaikov Kefu,

thank you very much for help!!!
The `flowinfo` is non zero on both Intel and Z, the `scope_id` is.
Let me test approach 1. I convert and compare the fields (operator==(), operator>=(), ... are need to have the same sequence of the structs in Intel and in Z, Radoslaw made an excellent study on this), this should be clear and will not trigger argues...

Reply:
Hi Kefu @tchaikov,
could it be approached in such a way?
[https://github.com/ibm-s390-cloud/ceph/commit/f5959a0687de05260c75f7e3bfdb402d7b7fefcc](https://github.com/ibm-s390-cloud/ceph/commit/f5959a0687de05260c75f7e3bfdb402d7b7fefcc)
Encoding/deconding is correct with the ceph_le types. Here only the sorting should make the same sequence on both LE and BE nodes.
Thank you!

Reply:
@aliakseimakarau and @rzarzynski replied at https://github.com/ibm-s390-cloud/ceph/commit/f5959a0687de05260c75f7e3bfdb402d7b7fefcc#r74163030